Rewrite custom form check backgrounds

mdo opened the pull request:
Previously, this was all just a `background-color` and `background-image`. When we restored the gradients though, we had two `background-image`s competing on the same element, causing rendering glitches. This refactors that code, creating a mixin to simplify things, so we can we easily use two `background-image`s (SVG icon and gradient) when `$enable-gradients` is set to true.

Fixes #24598.

---

#### Todo
- [x] Remove variable changes
- [x] Double check validation states of custom checks

mdo committed:
Rewrite custom form check backgrounds

Previously, this was all just a background-color and background-image. When we restored the gradients though, we had two background-images competing on the same element, causing rendering glitches. This refactors that code, creating a mixin to simplify things, so we can we easily use two background-images (SVG icon and gradient) when -gradients is set to true.

Fixes #24598

jonnysp commented:
@mdo thanks looks good. maybe change the validation status?
![1](https://user-images.githubusercontent.com/1651241/32479251-2ea99a20-c389-11e7-9778-fc09ad4a8c7f.JPG)
![3](https://user-images.githubusercontent.com/1651241/32479255-32765b70-c389-11e7-86e0-db1edbadbc07.JPG)
![2](https://user-images.githubusercontent.com/1651241/32479258-34d54e9e-c389-11e7-9d5b-d3bdca10d32f.JPG)
![4](https://user-images.githubusercontent.com/1651241/32479259-36016b9a-c389-11e7-94dc-df87284b461b.JPG)

andresgalante commented on scss/_variables.scss:
```diff
@@ -98,8 +98,8 @@ $yiq-text-light: $white !default;
 $enable-rounded:            true !default;
-$enable-shadows:            false !default;
-$enable-gradients:          false !default;
+$enable-shadows:            true !default;
```
Do we want to leave shadows and gradients on by default?

mdo replied:
Nope! Not yet at least :).

XhmikosR committed:
Merge branch 'v4-dev' into custom-form-check-bgs

mdo committed:
Merge branch 'v4-dev' into custom-form-check-bgs

mdo committed:
restore default vars

mdo committed:
Merge branch 'custom-form-check-bgs' of https://github.com/twbs/bootstrap into custom-form-check-bgs

mdo committed:
Revamp custom check and radio backgrounds

Instead of applying multiple background-image's to the same element, we're adding a new ::before pseudo-element to layer the background-images. Gradients go on the .custom-control-indicator while their icons go on the ::before element. This allows us to shave some bytes from when we compile and we previously needed to redeclare the background-image for the icon if you changed the gradient.

mdo committed:
remove now unused mixin

mdo commented:
Okay, decently sized update here:

I've had to change how we apply the `background-color` and multiple `background-image`s for custom checkboxes and radios. Previously, we had multiple comma-separated `background-image`s on the same `.custom-control-indicator` element. This presents a problem though for when you want to customize the background color or gradient without having to redeclare the icon, too. Now, the gradient and color are on `.custom-control-indicator` and the icon is on `.custom-control-indicator::before`.

I'll be adding a note to the migration docs for this here as well.

mdo committed:
mention change in migration docs

mdo commented:
@jonnysp With the changes I mentioned in my previous comment, you should be good to go with green and red background gradients on those custom forms.

jonnysp commented:
looks great, many thanks. can merge!

jonnysp reviewed (approved):
looks great, many thanks. can merge!
